fabtests/component/sock_test.c

```diff
@@ -0,0 +1,347 @@
+}
+
+static int init_select() 
+{
+	FD_ZERO(&select_set);
+	FD_SET(listen_sock, &select_set);
+	max_sock = listen_sock;
+
+	return 0;
+}
+
+static int time_select(int connections) 
+{
+	int i;
+	struct timeval timeout;
+	timeout.tv_sec = 0;
+	timeout.tv_usec = 0;
+
+	ft_start();
+	for (i = 0; i < connections; i++) {
+		FD_SET(clients[i], &select_set);
+
+		if (clients[i] > max_sock)
+			max_sock = clients[i];
+	}
+
+	select(max_sock + 1, &select_set, NULL, NULL, &timeout);
+	ft_stop();
+
+	printf("%i: %ld\n", connections, (end.tv_nsec + end.tv_sec * 1000000000) - 
+		(start.tv_nsec + start.tv_sec * 1000000000));
+	return 0;
+}
+
+static int init_poll(int connections)
+{
+	int i;
+	poll_set = malloc(sizeof(struct pollfd) * connections);
+	if (!poll_set)
+		return -1;
+
+	for (i = 0; i < connections; i++) {
+		poll_set[i].fd = clients[i];
+	}
+
+	return 0;
+}
+
+static int time_poll(int connections)
+{
+	ft_start();
+	poll(poll_set, POLLIN, 0);
+	ft_stop();
+
+	printf("%i: %ld\n", connections, (end.tv_nsec + end.tv_sec * 1000000000) - 
+		(start.tv_nsec + start.tv_sec * 1000000000));
+	return 0;
+}
+
+static int init_epoll(int connections)
+{
+	int i, ret;
+
+	efpd = epoll_create1(0);
+	events = malloc(sizeof(struct epoll_event) * connections);
+	if (!events)
+		return -1;
+
+	for (i = 0; i < connections; i++) {
+		events[i].events = EPOLLIN;
+		events[i].data.fd = clients[i];
+		ret = epoll_ctl(efpd, EPOLL_CTL_ADD, clients[i], &events[i]);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static int time_epoll(int connections)
+{
+	ft_start();
+	epoll_wait(EPOLLIN, events, connections, 0);
+	ft_stop();
+
+	printf("%i: %ld\n", connections, (end.tv_nsec + end.tv_sec * 1000000000) - 
+		(start.tv_nsec + start.tv_sec * 1000000000));
+
+	return 0;
+}
+
+static int init_resources(int connections, int (*poll_method) (int))
+{
+	int ret = 0;
+
+	if (poll_method == time_select)
+		ret = init_select();
+	else if (poll_method == time_poll)
+		ret = init_poll(connections);
+	else
+		ret = init_epoll(connections);
+
+	return ret;
+}
+
+static void free_resources()
+{
+	if (clients)
+		free(clients);
+	if (poll_set)
+		free(poll_set);
+	if (events)
+		free(events);
+}
+
+int main(int argc, char **argv)
+{
+	extern char *optarg;
+	int c, ret, connections, i;
+	char *mode;
+	int (*poll_method) (int);
+
+	opts = INIT_OPTS;
+	opts.options |= FT_OPT_SIZE;
+
+	while ((c = getopt(argc, argv, "n:m:" ADDR_OPTS)) != -1) {
+		switch (c) {
+		default:
+			ft_parse_addr_opts(c, optarg, &opts);
+			break;
+		case 'n':
+			connections = atoi(optarg);
+			break;
+		case 'm':
+			mode = optarg;
+			if (strcasecmp(mode, "select") == 0) {
+				printf("select mode\n");
+				poll_method = time_select;
+			} else if (strcasecmp(mode, "poll") == 0) {
+				printf("poll mode\n");
+				poll_method = time_poll;
+			} else if (strcasecmp(mode, "epoll")) {
+				printf("epoll mode\n");
+				poll_method = time_epoll;
+			} else {
+				printf("invalid mode");
+				return FI_EINVAL;
+			}
```

Comment:
Rather than running the test for 1 mode at a time, we can test all of them sequentially.